File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/keyboard/TouchCharInput.java
package net.kdt.pojavlaunch.customcontrols.keyboard;


import static android.content.Context.INPUT_METHOD_SERVICE;

import android.content.Context;
import android.text.Editable;
import android.text.Selection;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.inputmethod.InputMethodManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.arata.yukarilauncher.R;

/**
 * このクラスは、仮想キーボードを介してチャットで使用される文字を送信するためのものです
 */
public class TouchCharInput extends androidx.appcompat.widget.AppCompatEditText {
    public static final String TEXT_FILLER = "                              ";
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public TouchCharInput(@NonNull Context context) {
        this(context, null);
    }
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public TouchCharInput(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, R.attr.editTextStyle);
    }
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public TouchCharInput(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setup();
    }


    private boolean mIsDoingInternalChanges = false;
    private CharacterSenderStrategy mCharacterSender;

    /**
     * アプリ間を移動するとキーボードが無効になります。
     * そのため、オブジェクトを無効化します
     */
    @Override
    public void onWindowFocusChanged(boolean hasWindowFocus) {
        super.onWindowFocusChanged(hasWindowFocus);
        disable();
    }

    /**
     * バックキーをインターセプトしてフォーカスを無効にします
     * アクティビティの残りの部分には影響しません。
     */
    @Override
    public boolean onKeyPreIme(final int keyCode, final KeyEvent event) {
        if (event.getKeyCode() == KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_UP) {
            disable();
        }
        return super.onKeyPreIme(keyCode, event);
    }


    /**
     * 状態に応じてソフトキーボードのオン/オフを切り替えます
     */
    public void switchKeyboardState(){
        InputMethodManager imm = (InputMethodManager) getContext().getSystemService(INPUT_METHOD_SERVICE);
        // Allow, regardless of whether or not a hardware keyboard is declared
        if(hasFocus()){
            clear();
            disable();
        }else{
            enable();
            imm.showSoftInput(this, InputMethodManager.SHOW_IMPLICIT);
        }
    }


    /**
     * EditTextから残りの入力をクリアします
     * ゲーム内の入力には影響しません
     */
    public void clear(){
        mIsDoingInternalChanges = true;
        // Edit the Editable directly as it doesn't affect the state
        // of the TextView.
        Editable editable = getEditableText();
        editable.clear();
        // 点字スペース。キーボードの自動補完をトリガーしない
        editable.append(TEXT_FILLER);
        Selection.setSelection(editable, TEXT_FILLER.length());
        mIsDoingInternalChanges = false;
    }

    /**
     * 存在、フォーカス取得、テキスト入力の機能を回復します。
     */
    public void enable(){
        setEnabled(true);
        setFocusable(true);
        setVisibility(VISIBLE);
        requestFocus();
    }

    /**
     * 存在、フォーカス取得、テキスト入力の機能を失います。
     */
    public void disable(){
        clear();
        setVisibility(GONE);
        clearFocus();
        setEnabled(false);
        //setFocusable(false);
    }

    /**
     * Enterキーを送信します。
     */
    private void sendEnter(){
        mCharacterSender.sendEnter();
        clear();
    }

    /**
     * 使用する文字送信戦略を設定します。
     */
    public void setCharacterSender(CharacterSenderStrategy characterSender){
        mCharacterSender = characterSender;
    }

    /**
     * コンストラクタが呼び出されたときに実行する必要がある処理を行います。
     */
    private void setup(){
        // Using TextWatcher instead of overriding onTextChanged because some Huawei firmware
        // calls setText in constructor, causing havoc for our listener
        addTextChangedListener(new InputTextWatcher());
        setOnEditorActionListener((textView, i, keyEvent) -> {
            sendEnter();
            clear();
            disable();
            return false;
        });
        clear();
        disable();
    }
    private class InputTextWatcher implements android.text.TextWatcher {
/**
 * 「before Text Changed」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
        @Override
        public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

        }
        /**
         * 新しい文字を取得してゲームに送信します。
         * 文字数が少ない場合は、いくつか削除します。
         * テキストは常にクリーンアップされます。
         */
        @Override
        public void onTextChanged(CharSequence text, int start, int lengthBefore, int lengthAfter) {
            if(mIsDoingInternalChanges) return;
            if(mCharacterSender != null){
                for(int i=0; i < lengthBefore; ++i){
                    mCharacterSender.sendBackspace();
                }
                for(int i=start, count = 0; count < lengthAfter; ++i){
                    mCharacterSender.sendChar(text.charAt(i));
                    ++count;
                }
            }
        }
/**
 * 「after Text Changed」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
        @Override
        public void afterTextChanged(Editable editable) {
            if(mIsDoingInternalChanges) return;
            // Moved from onTextChanged because "It is an error to attempt to make changes to s from this callback."
            // reference: https://developer.android.com/reference/android/text/TextWatcher#onTextChanged(java.lang.CharSequence,%20int,%20int,%20int)
            if(editable.length() < 1) clear();
        }
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/keyboard/TouchCharInput.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols.keyboard

import android.content.Context
import android.text.Editable
import android.text.Selection
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.InputMethodManager
import com.arata.yukarilauncher.R

/**
 * タッチスクリーン用の文字入力ビュー。
 * AndroidのEditTextをベースに、Minecraftへの文字送信機能を提供します。
 * キーボード表示/非表示の切り替えや、入力文字のリアルタイム送信を行います。
 */
class TouchCharInput : androidx.appcompat.widget.AppCompatEditText {
    companion object {
        /** テキストバッファのフィラー文字列。 */
        const val TEXT_FILLER = "                              "
    }

    constructor(context: Context) : this(context, null)
    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, R.attr.editTextStyle)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        setup()
    }

    private var mIsDoingInternalChanges = false
    private var mCharacterSender: CharacterSenderStrategy? = null

    override fun onWindowFocusChanged(hasWindowFocus: Boolean) {
        super.onWindowFocusChanged(hasWindowFocus)
        disable()
    }

    override fun onKeyPreIme(keyCode: Int, event: KeyEvent): Boolean {
        if (event.keyCode == KeyEvent.KEYCODE_BACK && event.action == KeyEvent.ACTION_UP) {
            disable()
        }
        return super.onKeyPreIme(keyCode, event)
    }

    /** キーボードの表示状態を切り替えます。 */
    fun switchKeyboardState() {
        val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        if (hasFocus()) {
            clear()
            disable()
        } else {
            enable()
            imm.showSoftInput(this, InputMethodManager.SHOW_IMPLICIT)
        }
    }

    /** テキストバッファをクリアします。 */
    fun clear() {
        mIsDoingInternalChanges = true
        val editable = editableText
        editable.clear()
        editable.append(TEXT_FILLER)
        Selection.setSelection(editable, TEXT_FILLER.length)
        mIsDoingInternalChanges = false
    }

    /** 入力を有効にします。 */
    fun enable() {
        isEnabled = true
        isFocusable = true
        visibility = View.VISIBLE
        requestFocus()
    }

    /** 入力を無効にします。 */
    fun disable() {
        clear()
        visibility = View.GONE
        clearFocus()
        isEnabled = false
    }

    /** エンターキーを送信します。 */
    private fun sendEnter() {
        mCharacterSender?.sendEnter()
        clear()
    }

    /** @param characterSender 文字送信戦略を設定 */
    fun setCharacterSender(characterSender: CharacterSenderStrategy?) {
        mCharacterSender = characterSender
    }

    /** 初期設定を行います。 */
    private fun setup() {
        addTextChangedListener(InputTextWatcher())
        setOnEditorActionListener { _, _, _ ->
            sendEnter()
            clear()
            disable()
            false
        }
        clear()
        disable()
    }

    /** 入力テキストの変更を監視し、Minecraftに文字を送信する内部リスナークラス。 */
    private inner class InputTextWatcher : android.text.TextWatcher {
        override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}

        override fun onTextChanged(text: CharSequence, start: Int, lengthBefore: Int, lengthAfter: Int) {
            if (mIsDoingInternalChanges) return
            mCharacterSender?.let { sender ->
                for (i in 0 until lengthBefore) {
                    sender.sendBackspace()
                }
                var count = 0
                var i = start
                while (count < lengthAfter) {
                    sender.sendChar(text[i])
                    count++
                    i++
                }
            }
        }

        override fun afterTextChanged(editable: Editable) {
            if (mIsDoingInternalChanges) return
            if (editable.length < 1) clear()
        }
    }
}
